internal/osutils/lockfile/lock_test.go

```diff
@@ -0,0 +1,198 @@
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func buildTestExecutable(t *testing.T, dir string) string {
+	root, err := environment.GetRootPath()
+	require.NoError(t, err)
+	lockerExe := filepath.Join(dir, "locker")
+	if runtime.GOOS == "windows" {
+		lockerExe += ".exe"
+	}
+
+	cmd := exec.Command(
+		"go", "build", "-o", lockerExe,
+		filepath.Join(root, "internal", "osutils", "testdata", "locker"),
+	)
+	err = cmd.Run()
+	require.NoError(t, err)
+
+	return lockerExe
+}
+
+func Test_acquirePidLockProcesses(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "")
+	require.NoError(t, err)
+	defer os.RemoveAll(tmpDir)
+
+	// build the test process that acquires a lock
+	lockerExe := buildTestExecutable(t, tmpDir)
+
+	t.Run("locked in other process", func(tt *testing.T) {
+		lockFile := filepath.Join(tmpDir, "locked-")
+
+		lockCmd := exec.Command(lockerExe, lockFile)
+		lockCmd = prepLockCmd(lockCmd)
+
+		stdout, err := lockCmd.StdoutPipe()
+		require.NoError(tt, err)
+		err = lockCmd.Start()
+		require.NoError(tt, err)
+
+		// wait for command to block
+		buf := make([]byte, 6)
+		n, err := stdout.Read(buf)
+		require.NoError(tt, err)
+		require.Equal(tt, 6, n)
+		assert.Equal(tt, "LOCKED", string(buf))
+
+		pl, err := NewLock(lockFile)
+		require.NoError(tt, err)
+
+		// trying to acquire the lock in this process should fail
+		err = pl.TryLock()
+		require.Error(tt, err)
+		alreadyErr := &AlreadyLockedError{}
+		assert.True(tt, errors.As(err, &alreadyErr))
+
+		err = pl.Close()
+		require.NoError(tt, err)
+
+		// stopping the other process
+		interruptProcess(tt, lockCmd.Process)
+
+		// waiting for the process to finish without error
+		err = lockCmd.Wait()
+		require.NoError(tt, err)
+		assert.True(tt, lockCmd.ProcessState.Exited())
+		assert.Equal(tt, 0, lockCmd.ProcessState.ExitCode())
+	})
+
+	t.Run("stress-test", func(tt *testing.T) {
+		// stress tests runs numProcesses in parallel, and only one should get the lock
+		numProcesses := 10
+
+		lockFile := filepath.Join(tmpDir, "stress-test")
+
+		done := make(chan string, numProcesses+1)
+		defer close(done)
+		var wg sync.WaitGroup
+		defer wg.Wait()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		for i := 0; i < numProcesses; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				lockCmd := exec.Command(lockerExe, lockFile)
+				lockCmd = prepLockCmd(lockCmd)
+				stdout, err := lockCmd.StdoutPipe()
+				require.NoError(tt, err)
+				err = lockCmd.Start()
+				require.NoError(tt, err)
+
+				// wait for command to block
+				buf := make([]byte, 6)
+				n, err := stdout.Read(buf)
+				require.NoError(tt, err)
+				require.Equal(tt, 6, n)
+
+				sb := string(buf[:n])
+				if sb == "DENIED" {
+					done <- "DENIED"
+					return
+				}
+				assert.Equal(tt, "LOCKED", string(buf[:6]))
+
+				done <- "LOCKED"
+
+				// wait for the signal to kill process and to release the lock
+				<-ctx.Done()
```

Comment:
I suggest placing the select/timeout here and using the waitgroup where the timeout is now. The deferred wait call significantly disconcerting.

Reply:
Though, if it's the simplest way for you to achieve your goal, no worries.

Reply:
I don't remember why I added the second go function.  I removed it...